File: app/src/main/java/com/sinthoras/randograf/cards/CardTwoStructures.java
package com.sinthoras.randograf.cards;

import android.content.res.Resources;

import com.sinthoras.randograf.Colors;
import com.sinthoras.randograf.R;
import com.sinthoras.randograf.structure.Structure;
import com.sinthoras.randograf.structure.StructureGenerator;

import java.util.Random;

public abstract class CardTwoStructures extends Card {

    private final Structure structureA;
    private final Structure structureB;

    protected final Random random = new Random();

    protected CardTwoStructures(Colors colorA) {
        structureA = StructureGenerator.generateStructure(randomSizeForStrucutreA()).withColor(colorA);
        structureB = generateStructureB(structureA);
    }

    private int randomSizeForStrucutreA() {
        return 4 + random.nextInt() % 2;
    }

    protected abstract Structure generateStructureB(Structure structureA);

    @Override
    public void paint() {
        // TODO
    }

    @Override
    public String getTitle() {
        return Resources.getSystem().getString(R.string.card_title_two_options);
    }
}

File: app/src/main/java/com/sinthoras/randograf/cards/CardTwoStructures.kt
package com.sinthoras.randograf.cards

import com.sinthoras.randograf.BlockColors
import com.sinthoras.randograf.R
import com.sinthoras.randograf.Random.getRandomFromTo
import com.sinthoras.randograf.structure.Structure
import com.sinthoras.randograf.structure.StructureGenerator

class CardTwoStructures(
    color: BlockColors
) : Card() {

    private val structureSmall: Structure
    private val structureLarge: Structure

    fun getStructureSmall() = structureSmall

    fun getStructureLarge() = structureLarge

    override fun getTitle(): Int {
        return R.string.card_title_two_structures
    }

    init {
        structureLarge = StructureGenerator.generateStructure(getRandomFromTo(4, 5)).withColor(color)
        structureSmall = StructureGenerator.generateStructure(getSizeForSmallStructure(structureLarge)).withColor(color)
    }

    override fun getDuration() = 1

    private fun getSizeForSmallStructure(structureLarge: Structure): Int {
        if(structureLarge.size() <= 4) {
            return 2;
        }
        return getRandomFromTo(2, 3);
    }
}
